### PR TITLE
ChangeId for the client

### PR DESCRIPTION
**Pull request by FinnStutzenstein**

### CODE REVIEW

**Comment by FinnStutzenstein** [comment]
@ostcar See emai. If I request change x+1 on reload, the server gives an error.

**Comment by FinnStutzenstein** [comment]
@jsaalfeld Yes, offline mode is not the scope here. This should be a drop in with the change id enabled. The offline stuff are (if exist) stubs; These are places where (possibly) logic joins together.

I'll remove some of the console.log. Emanuel wants to have some left for presentation; then they can be removed

What queue do you mean? for requests during offline mode? See above ;)

The "only take effect after you execute the same command twice" I'll check..

File: client/src/app/core/services/auth.service.ts [hunk 1]
```diff
@@ -32,9 +31,10 @@ export class AuthService extends OpenSlidesComponent {
      * @param operator who is using OpenSlides
```
Comment (jsaalfeld) [inline client/src/app/core/services/auth.service.ts]: paramters in documentation are outdated

File: client/src/app/core/services/auth.service.ts [hunk 2]
```diff
@@ -49,17 +49,14 @@ export class AuthService extends OpenSlidesComponent {
      * @param username
      * @param password
```
Comment (jsaalfeld) [inline client/src/app/core/services/auth.service.ts]: return paremeter is missing documentation

File: client/src/app/core/services/autoupdate.service.ts [hunk 3]
```diff
@@ -43,12 +54,12 @@ export class AutoupdateService extends OpenSlidesComponent {
      * @param websocketService
```
Comment (jsaalfeld) [inline client/src/app/core/services/autoupdate.service.ts]: some parameters are missing in the documentation

File: client/src/app/core/services/autoupdate.service.ts [hunk 4]
```diff
@@ -63,30 +74,80 @@ export class AutoupdateService extends OpenSlidesComponent {
      *
      * Saves models in DataStore.
```
Comment (jsaalfeld) [inline client/src/app/core/services/autoupdate.service.ts]: parameter is missing in the documentation

File: client/src/app/core/services/autoupdate.service.ts [hunk 5]
```diff
@@ -63,30 +74,80 @@ export class AutoupdateService extends OpenSlidesComponent {
...
+            // autoupdate fully in the future. we are missing something!
+            this.requestChanges();
+        }
+    }
+
+    /**
+     * Creates baseModels for each plain object
+     * @param collection The collection all models have to be from.
+     * @param models All models that should be mapped to BaseModels
+     */
```
Comment (jsaalfeld) [inline client/src/app/core/services/autoupdate.service.ts]: return statement is missing in the documentation

File: client/src/app/core/services/data-store.service.ts [hunk 6]
```diff
@@ -108,37 +108,43 @@ export class DataStoreService {
 
     /**
      * Empty constructor for dataStore
-     * @param cacheService use CacheService to cache the DataStore.
+     * @param storageService use CacheService to cache the DataStore.
```
Comment (jsaalfeld) [inline client/src/app/core/services/data-store.service.ts]: new parameters are missing in the documentation

File: client/src/app/core/services/data-store.service.ts [hunk 7]
```diff
@@ -108,37 +108,43 @@ export class DataStoreService {
...
      */
-    public constructor(private cacheService: CacheService, private modelMapper: CollectionStringModelMapperService) {}
+    public constructor(
+        private storageService: StorageService,
+        private modelMapper: CollectionStringModelMapperService
+    ) {}
 
     /**
      * Gets the DataStore from cache and instantiate all models out of the serialized version.
```
Comment (jsaalfeld) [inline client/src/app/core/services/data-store.service.ts]: return statement is missing in the documentation

File: client/src/app/core/services/data-store.service.ts [hunk 8]
```diff
@@ -163,15 +169,19 @@ export class DataStoreService {
      * Clears the complete DataStore and Cache.
      * @param callback
```
Comment (jsaalfeld) [inline client/src/app/core/services/data-store.service.ts]: the parameter is outdated

File: client/src/app/core/services/data-store.service.ts [hunk 9]
```diff
@@ -163,15 +169,19 @@ export class DataStoreService {
...
-            this.cacheService.remove(DataStoreService.cachePrefix + 'maxChangeId', callback);
-        });
+        await this.storageService.remove(DataStoreService.cachePrefix + 'DS');
+        await this.storageService.remove(DataStoreService.cachePrefix + 'maxChangeId');
     }
 
+    /**
+     * Returns the collection _string_ based on the model given. If a string is given, it's just returned.
+     * @param collectionType Either a Model constructor or a string.
+     */
```
Comment (jsaalfeld) [inline client/src/app/core/services/data-store.service.ts]: return value is missing in the documentation

File: client/src/app/core/services/data-store.service.ts [hunk 10]
```diff
@@ -262,62 +272,78 @@ export class DataStoreService {
...
-        this.storeToCache(changeId);
+        if (changeId) {
+            await this.flushToStorage(changeId);
+        }
     }
 
     /**
      * removes one or multiple models from dataStore.
      *
      * @param Type The desired BaseModel type to be read from the datastore
```
Comment (jsaalfeld) [inline client/src/app/core/services/data-store.service.ts]: some of the parameters are outdated

File: client/src/app/core/services/offline.service.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import { Injectable } from '@angular/core';
+
+import { OpenSlidesComponent } from 'app/openslides.component';
+import { DataStoreService } from './data-store.service';
+import { WhoAmIResponse } from './operator.service';
+
+/**
+ * @ignore
```
Comment (jsaalfeld) [inline client/src/app/core/services/offline.service.ts]: why should this be ignored?

File: client/src/app/core/services/offline.service.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+export class OfflineService extends OpenSlidesComponent {
+    private _offline = false;
+
+    public get offline(): boolean {
+        return this._offline;
+    }
+
+    /**
+     * Constructor to create the AutoupdateService. Calls the constructor of the parent class.
+     * @param websocketService
```
Comment (jsaalfeld) [inline client/src/app/core/services/offline.service.ts]: the parameter is outdated

File: client/src/app/core/services/openslides.service.ts [hunk 13]
```diff
@@ -22,14 +22,14 @@ export class OpenSlidesService extends OpenSlidesComponent {
 
     /**
      * Constructor to create the NotifyService. Registers itself to the WebsocketService.
-     * @param cacheService
+     * @param storageService
      * @param operator
      * @param websocketService
      * @param router
      * @param autoupdateService
```
Comment (jsaalfeld) [inline client/src/app/core/services/openslides.service.ts]: the `DS` parameter is missing in the documentation

File: client/src/app/core/services/operator.service.ts [hunk 14]
```diff
@@ -75,7 +75,7 @@ export class OperatorService extends OpenSlidesComponent {
     /**
      * @param http HttpClient
```
Comment (jsaalfeld) [inline client/src/app/core/services/operator.service.ts]: the documentation for this function is lacking some details

File: client/src/app/core/services/operator.service.ts [hunk 15]
```diff
@@ -98,15 +98,19 @@ export class OperatorService extends OpenSlidesComponent {
     /**
      * Calls `/apps/users/whoami` to find out the real operator.
```
Comment (jsaalfeld) [inline client/src/app/core/services/operator.service.ts]: The return statement is missing in the documentaiton

File: client/src/app/core/services/storage.service.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+     * Constructor to create the CacheService. Needs the localStorage service.
+     * @param localStorage
+     */
+    public constructor(private localStorage: LocalStorage) {}
+
+    /**
+     * Sets the item into the store asynchronously.
+     * @param key
+     * @param item
+     * @param callback An optional callback that is called on success
```
Comment (jsaalfeld) [inline client/src/app/core/services/storage.service.ts]: `callback` parameter seems to be outdated

File: client/src/app/core/services/storage.service.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+     * @param key The key to get the value from
+     */
+    public async get<T>(key: string): Promise<T> {
+        return await this.localStorage.getItem<T>(key).toPromise();
+    }
+
+    /**
+     * Remove the key from the store.
+     * @param key The key to remove the value from
+     * @param callback An optional callback that is called on success
```
Comment (jsaalfeld) [inline client/src/app/core/services/storage.service.ts]: `callback` parameter seems to be outdated

File: client/src/app/core/services/storage.service.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public async remove(key: string): Promise<void> {
+        if (!(await this.localStorage.removeItem(key).toPromise())) {
+            throw new Error('Could not delete the item.');
+        }
+    }
+
+    /**
+     * Clear the whole cache
+     * @param callback An optional callback that is called on success
```
Comment (jsaalfeld) [inline client/src/app/core/services/storage.service.ts]: `callback` paramter seems to be outdated

File: client/src/app/core/services/websocket.service.ts [hunk 19]
```diff
@@ -69,6 +69,8 @@ export class WebsocketService {
      */
     private subjects: { [type: string]: Subject<any> } = {};
 
+    private retry = false;
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: a documentation for this variable would be nice

File: client/src/app/core/services/websocket.service.ts [hunk 20]
```diff
@@ -69,6 +69,8 @@ export class WebsocketService {
...
+
     /**
      * Constructor that handles the router
      * @param router the URL Router
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: the documentation is lacking some parameters

File: client/src/app/core/services/websocket.service.ts [hunk 21]
```diff
@@ -85,34 +87,48 @@ export class WebsocketService {
      *
      * Uses NgZone to let all callbacks run in the angular context.
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: you are missing some parameters in the documentation

File: client/src/app/core/services/websocket.service.ts [hunk 22]
```diff
@@ -85,34 +87,48 @@ export class WebsocketService {
...
+        options: {
+            changeId?: number;
+            enableAutoupdates?: boolean;
+        } = {}
+    ): void {
         if (this.websocket) {
-            return;
+            this.close();
         }
+        console.log('connect', options);
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: is this `console.log` needed?

File: client/src/app/core/services/websocket.service.ts [hunk 23]
```diff
@@ -85,34 +87,48 @@ export class WebsocketService {
...
+
+        if (options.changeId !== undefined) {
+            queryParams.change_id = options.changeId;
+        }
 
         // Create the websocket
         const socketProtocol = this.getWebSocketProtocol();
         const socketServer = window.location.hostname + ':' + window.location.port;
         const socketPath = this.getWebSocketPath(queryParams);
+        console.log('connect to', socketProtocol + socketServer + socketPath);
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: is this `console.log` needed?

File: client/src/app/core/services/websocket.service.ts [hunk 24]
```diff
@@ -85,34 +87,48 @@ export class WebsocketService {
...
         this.websocket = new WebSocket(socketProtocol + socketServer + socketPath);
 
         // connection established. If this connect attept was a retry,
         // The error notice will be removed and the reconnectSubject is published.
         this.websocket.onopen = (event: Event) => {
             this.zone.run(() => {
-                if (retry) {
+                console.log('onopen', options);
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: is this `console.log` needed?

File: client/src/app/core/services/websocket.service.ts [hunk 25]
```diff
@@ -137,6 +153,7 @@ export class WebsocketService {
         this.websocket.onclose = (event: CloseEvent) => {
             this.zone.run(() => {
                 this.websocket = null;
+                console.log('ws close', event.code);
```
Comment (jsaalfeld) [inline client/src/app/core/services/websocket.service.ts]: is this `console.log` needed?

**Pull request merged**